File: 2411-spiral-matrix-iv/2411-spiral-matrix-iv.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func spiralMatrix(m int, n int, head *ListNode) [][]int {
    ans := make([][]int, m)
    for i := range ans {
        ans[i] = make([]int, n)
        for j := range ans[i] {
            ans[i][j] = -1
        }
    }

    rowStart := 0
    rowEnd := m - 1
    colStart := 0
    colEnd := n - 1

    for head != nil {
        for col := colStart; col <= colEnd && head != nil; col++ {
            ans[rowStart][col] = head.Val
            head = head.Next
        }
        rowStart++

        for row := rowStart; row <= rowEnd && head != nil; row++ {
            ans[row][colEnd] = head.Val
            head = head.Next
        }
        colEnd--

        for col := colEnd; col >= colStart && head != nil; col-- {
            ans[rowEnd][col] = head.Val
            head = head.Next
        }
        rowEnd--

        for row := rowEnd; row >= rowStart && head != nil; row-- {
            ans[row][colStart] = head.Val
            head = head.Next
        }
        colStart++
    }

    return ans
}

File: 2411-spiral-matrix-iv/2411-spiral-matrix-iv.cpp
#pragma GCC optimize("03", "unroll-loops");
static const int __ = [](){
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 0;
}();

class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        vector<vector<int>> ans(m, vector<int>(n, -1));
        int rowStart = 0, rowEnd = m - 1;
        int colStart = 0, colEnd = n - 1;
        
        while (head && rowStart <= rowEnd && colStart <= colEnd) {
            for (int col = colStart; col <= colEnd && head; ++col) {
                ans[rowStart][col] = head->val;
                head = head->next;
            }
            rowStart++;
            
            for (int row = rowStart; row <= rowEnd && head; ++row) {
                ans[row][colEnd] = head->val;
                head = head->next;
            }
            colEnd--;
            
            for (int col = colEnd; col >= colStart && head; --col) {
                ans[rowEnd][col] = head->val;
                head = head->next;
            }
            rowEnd--;
            
            for (int row = rowEnd; row >= rowStart && head; --row) {
                ans[row][colStart] = head->val;
                head = head->next;
            }
            colStart++;
        }
        
        return ans;
    }
};